io-sim/src/Control/Monad/IOSim/Internal.hs

```diff
@@ -296,11 +296,22 @@ schedule !thread@Thread{
           -- An unhandled exception in any other thread terminates the thread
           let reason | ThrowSelf <- thrower = FinishedNormally
                      | otherwise            = FinishedDied
-          !trace <- deschedule (Terminated reason) thread simstate
+          !trace <- deschedule (Terminated reason) thread simstate { timers = timers' }
           return $ SimTrace time tid tlbl (EventThrow e)
                  $ SimTrace time tid tlbl (EventThreadUnhandled e)
                  $ SimTrace time tid tlbl (EventDeschedule $ Terminated reason)
                  $ trace
+      where
+        -- When we throw an exception we need to cancel thread delay action
+        -- which could be blocking the current thread.  This is important for
+        -- `TimeoutException` or any other asynchronous exceptions.
+        timers' = PSQ.fromList
+                . filter (\(_,_,info) -> case info of
+                            TimerThreadDelay tid' _
+                              -> tid' /= tid
+                            _ -> True)
+                . PSQ.toAscList
```

Comment:
This is quite expensive. It's O(n) in the set of all timers.

Here what we have is the thread id. We could look up the thread id to get the `Thread`. We currently track that threads are blocked as a `Bool` but we could make that structured to track the reason for blocking. The blocking on a timer would include the timer id. This would allow us to get at the timer id and then cancel the timer. I think this is how its done in the GHC RTS.

Reply:
Fixed in #86.